### PR TITLE
React web update

### PR DESCRIPTION
**Pull request by ErisMik**

### CODE REVIEW

File: binding/react/src/use_rhino.ts [hunk 1]
```diff
@@ -9,276 +9,108 @@
...
+    } catch (e: any) {
+      setError(e.toString());
     }
+  }, [rhino, accessKey, context, model, options]);
 
-    if (rhinoHookArgs === null || rhinoHookArgs === undefined) {
-      return (): void => {
-        /* NOOP */
-      };
+  const pushToTalk = useCallback(async (): Promise<void> => {
```
Comment (mrrostam) [inline binding/react/src/use_rhino.ts:80]: is it ok if someone uses pushToTalk twice before finalizing the first inference? if I'm not mistaken we subscribe the engine twice and that could cause some issues, right?

Comment (ErisMik) [reply binding/react/src/use_rhino.ts:80]: Excellent point, it will cause issues so I added in a check

**Review by mrrostam — APPROVED** [review]

**Review by ksyeo1010 — APPROVED** [review]

File: binding/react/src/use_rhino.ts [hunk 2]
```diff
@@ -9,277 +9,114 @@
...
-  const [isError, setIsError] = useState<boolean | null>(null);
-  const [isListening, setIsListening] = useState(false);
-  const [isLoaded, setIsLoaded] = useState(false);
-  const [isTalking, setIsTalking] = useState(false);
+  error: string | null;
+  init: (
+    accessKey: string,
+    context: RhinoContext,
+    model: RhinoModel,
+    options: RhinoOptions
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:33]:
```suggestion
    options?: RhinoOptions
```

File: binding/react/src/use_rhino.ts [hunk 3]
```diff
@@ -9,277 +9,114 @@
...
+  );
+
+  const init = useCallback(
+    async (
+      accessKey: string,
+      context: RhinoContext,
+      model: RhinoModel,
+      options: RhinoOptions = {}
+    ): Promise<void> => {
+      try {
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:65]: check options.processErrorCallback

File: demo/react/src/VoiceWidget.jsx [hunk 4]
```diff
@@ -69,50 +43,37 @@ export default function VoiceWidget() {
...
           Start Rhino
         </button>
       </h3>
-      <h3>Dynamic Import Loaded: {JSON.stringify(isChunkLoaded)}</h3>
       <h3>Rhino Loaded: {JSON.stringify(isLoaded)}</h3>
       <h3>Listening: {JSON.stringify(isListening)}</h3>
-      <h3>Error: {JSON.stringify(isError)}</h3>
-      {isError && accessKey && (
-        <p className="error-message">{JSON.stringify(errorMessage)}</p>
+      <h3>Error: {JSON.stringify(error !== null)}</h3>
```
Comment (ksyeo1010) [inline demo/react/src/VoiceWidget.jsx:56]: add Loaded as well

File: binding/react/src/use_rhino.ts [hunk 5]
```diff
@@ -9,277 +9,114 @@
...
-    // ".... we know our data structure is relatively shallow, doesn't have cycles,
-    // and is easily serializable ... doesn't have functions or weird objects like Dates.
-    // ... it's acceptable to pass [JSON.stringify(variables)] as a dependency."
-    JSON.stringify(rhinoHookArgs),
-  ]);
+    setIsListening(false);
+  }, []);
 
   return {
+    inference,
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:113]: add a clean up function:
```
useEffect(() => {
  return () => {
    if (rhinoRef.current) ...
  }
}, [])
```

**Review by ksyeo1010 — CHANGES_REQUESTED** [review]

File: demo/react/src/VoiceWidget.jsx (hunk 4 above)
Comment (ErisMik) [reply demo/react/src/VoiceWidget.jsx:56]: It is there :)

File: binding/react/src/use_rhino.ts [hunk 6]
```diff
@@ -9,277 +9,129 @@
...
+        if (!rhinoRef.current) {
+          rhinoRef.current = await RhinoWorker.create(
+            accessKey,
+            context,
+            inferenceCallback,
+            model,
+            { ...options, processErrorCallback: errorCallback }
+          );
+          setContextInfo(rhinoRef.current.contextInfo);
+          setIsLoaded(true);
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:86]:
```suggestion
          setIsLoaded(true);
          setError(null);
```

File: binding/react/src/use_rhino.ts [hunk 7]
```diff
@@ -9,277 +9,129 @@
...
+    },
+    [inferenceCallback]
+  );
+
+  const process = useCallback(async (): Promise<void> => {
+    try {
+      if (rhinoRef.current && !isListening) {
+        rhinoRef.current.reset();
+        await WebVoiceProcessor.subscribe(rhinoRef.current);
+        setIsListening(true);
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:100]:
```suggestion
        setIsListening(true);
        setError(null);
```

File: binding/react/src/use_rhino.ts [hunk 8]
```diff
@@ -9,277 +9,129 @@
...
+  }, [isListening]);
+
+  const release = useCallback(async (): Promise<void> => {
+    try {
+      if (rhinoRef.current) {
+        await WebVoiceProcessor.unsubscribe(rhinoRef.current);
+        rhinoRef.current.release();
+        rhinoRef.current.terminate();
+        rhinoRef.current = null;
+        setIsListening(false);
```
Comment (ksyeo1010) [inline binding/react/src/use_rhino.ts:114]:
```suggestion
        setIsListening(false);
        setError(null);
```

**Review by ksyeo1010 — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -1396,57 +1396,47 @@ public pushToTalk() {
 #### React
 
 ```console
-yarn add @picovoice/rhino-web-react @picovoice/rhino-web-en-worker
+yarn add @picovoice/rhino-react @picovoice/web-voice-processor
```
Comment (laves) [inline README.md:1399]: don't think we need to tell the user to explicitly import web voice processor

File: binding/react/README.md [hunk 10]
```diff
@@ -1,194 +1,210 @@
-# rhino-web-react
+# Rhino Binding for Web
```
Comment (laves) [inline binding/react/README.md:1]:
```suggestion
# Rhino Binding for React
```

File: binding/react/README.md [hunk 11]
```diff
@@ -1,194 +1,210 @@
...
 
-## Compatibility
+## Installation
 
-The Picovoice SDKs for Web are powered by WebAssembly (WASM), the Web Audio API, and Web Workers.
+Using `yarn`:
 
-All modern browsers (Chrome/Edge/Opera, Firefox, Safari) are supported, including on mobile. Internet Explorer is _not_ supported.
+```console
+yarn add @picovoice/rhino-react @picovoice/web-voice-processor
```
Comment (laves) [inline binding/react/README.md:49]: don't need to add wvp, it's a dependency of `rhino-react`

File: binding/react/README.md [hunk 12]
```diff
@@ -1,194 +1,210 @@
...
+    rhinoModel,
+    options
+  )
+}
+```
+
+The `init` process is asyncronous. Once complete, the `isLoaded` flag will be set to `true`.
 
-  const [workerChunk, setWorkerChunk] = useState({ factory: null });
+### Process Audio Frames in Worker Thread
```
Comment (laves) [inline binding/react/README.md:172]:
```suggestion
### Process Audio Frames
```

File: binding/react/README.md [hunk 13]
```diff
@@ -1,194 +1,210 @@
...
+## Non-English Languages
+
+In order to detect non-English inferences you need to use the corresponding model file (`.pv`). The model files for all
+supported languages are available [here](https://github.com/Picovoice/rhino/tree/master/lib/common).
 
-Custom contexts are generated using [Picovoice Console](https://console.picovoice.ai/). They are trained from text using transfer learning into bespoke Rhino context files with a `.rhn` extension. The target platform is WebAssembly (WASM), as that is what backs the React library.
+## Demo
 
-The `.zip` file contains a `.rhn` file and a `_b64.txt` file which contains the binary model encoded with Base64. Provide the base64 encoded string as an argument to Rhino as in the above example. You may wish to store the base64 string in a separate JavaScript file and `export` it to keep your application code separate.
+For example usage refer to our [Web demo application](https://github.com/Picovoice/rhino/tree/master/demo/web).
```
Comment (laves) [inline binding/react/README.md:210]:
```suggestion
For example usage, refer to our [React demo application](https://github.com/Picovoice/rhino/tree/master/demo/react).
```

File: binding/react/src/use_rhino.ts [hunk 14]
```diff
@@ -9,277 +9,132 @@
...
-  pause: () => void;
-  /** A method to put Rhino in an active isTalking state */
-  pushToTalk: () => void;
-} {
-  const [errorMessage, setErrorMessage] = useState<string | null>(null);
-  const [isError, setIsError] = useState<boolean | null>(null);
-  const [isListening, setIsListening] = useState(false);
-  const [isLoaded, setIsLoaded] = useState(false);
-  const [isTalking, setIsTalking] = useState(false);
+  error: Error | string | null;
```
Comment (laves) [inline binding/react/src/use_rhino.ts:28]: are we ever passing an `Error`? I thought we just did string

File: binding/react/src/use_rhino.ts [hunk 15]
```diff
@@ -9,277 +9,132 @@
...
+    async (
+      accessKey: string,
+      context: RhinoContext,
+      model: RhinoModel,
+      options: RhinoOptions = {}
+    ): Promise<void> => {
+      if (options.processErrorCallback) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          "'processErrorCallback' options is not supported, use the 'error' hook variable instead."
```
Comment (laves) [inline binding/react/src/use_rhino.ts:72]: Kwangsoo and I used: 
console.warn("'processErrorCallback' is only supported in the Rhino Web SDK. Use the 'error' state to monitor for errors in the React SDK.");

File: binding/react/src/use_rhino.ts [hunk 16]
```diff
@@ -9,277 +9,132 @@
...
+      } catch (e: any) {
+        errorCallback(e.toString());
+      }
+    },
+    [inferenceCallback]
+  );
+
+  const process = useCallback(async (): Promise<void> => {
+    try {
+      if (rhinoRef.current && !isListening) {
```
Comment (laves) [inline binding/react/src/use_rhino.ts:98]: Might be good to check for state problems (fail loud):
```typescript
if (!rhinoRef.current) {
  setError('Rhino has not been initialized or has been released');
  return;
}
```

File: binding/react/src/use_rhino.ts [hunk 17]
```diff
@@ -9,277 +9,132 @@
...
-      } catch (error) {
-        rhnWorker.postMessage({ command: 'release' });
-        throw error;
+  }, [isListening]);
+
+  const release = useCallback(async (): Promise<void> => {
+    try {
+      if (rhinoRef.current) {
+        await WebVoiceProcessor.unsubscribe(rhinoRef.current);
+        rhinoRef.current.release();
```
Comment (laves) [inline binding/react/src/use_rhino.ts:113]: according to Kwangsoo, we don't need to call release, just terminate

File: demo/react/src/VoiceWidget.jsx [hunk 18]
```diff
@@ -69,50 +43,37 @@ export default function VoiceWidget() {
...
             type="text"
             name="accessKey"
-            onChange={(value) => setInputValue(value.target.value)}
+            onChange={(value) => setAccessKey(value.target.value)}
             disabled={isLoaded}
           />
         </label>
-        <button className="start-button" onClick={() => setAccessKey(inputValue)} disabled={isLoaded}>
+        <button className="start-button" onClick={() => rhnInit()} disabled={isLoaded || accessKey.length === 0}>
           Start Rhino
```
Comment (laves) [inline demo/react/src/VoiceWidget.jsx:51]: Init Rhino

**Review by laves — CHANGES_REQUESTED** [review]

**Review by laves — APPROVED** [review]

**Pull request merged**